### PR TITLE
feat: Add tools

### PR DESCRIPTION
**Pull request by hannasoderstrom1**
### This PR:

Added: Module with Sentieon tools

Mapping:
BWA mem
Dedup - remove duplicate reads
Realigner - indel realignment

QualCal - calculate recalibration table

Variant calling:
DNAScope - Germline SNV/structural variant calling
DNAScope modelfilter - Modify dnascope vcf by adding the MLrejected filter
TNScope - Somatic SNV/structural variant calling using matched normal sample
TNScope modelfilter -  Modify tnscope vcf by adding the MLrejected filter

### Review and tests: 

Have test run the rules and gotten the correct outputs.

### To do / Issues

Add rule for TNScope tumor only

Filtering steps (using bcftools) to filter vcf outputs from both DNAScope and TNScope. Should the filtering steps be in another module? 

Change some outputs to temp

For now, local singularity is used since you need a license for Sentieon

The realigned bam files should be able to be used as input in other hydra rules as well, for example manta, instead of bam from hydra alignment module. Want to be able to choose sentieon mapping or mapping by hydra alignment module.

### CODE REVIEW

File: workflow/rules/sentieon.smk [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        threads=config.get("sentieon", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("sentieon", {}).get("time", config["default_resources"]["time"]),
+    container:
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Indel realignment of bam file {input.bam} using Sentieon realigner"
+    shell:
+        "{params.sentieon} driver -t {threads} -r {params.reference} -i {input.bam} --algo Realigner -k {params.mills} {output}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]: New line for each parameter

File: workflow/rules/sentieon.smk [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        threads=config.get("sentieon", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("sentieon", {}).get("time", config["default_resources"]["time"]),
+    container:
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Calculate recalibration table of {input.bam} using Sentieon QualCal algorithm"
+    shell:
+        "{params.sentieon} driver -t {threads} -r {params.reference} -i {input.bam} --algo QualCal -k {params.mills} -k {params.dbsnp} {output}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]: New line for each parameter

File: workflow/rules/sentieon.smk [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    container:
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Call germline SNVs and structural variants in {input.bam} using Sentieon DNAScope"
+    shell:
+        "{params.sentieon} driver -t {threads} -r {params.reference} "
+            "-i {input.bam} --algo DNAscope -d {params.dbsnp} "
+            "--var_type snp,indel --model {params.model} {params.callsettings} {output.dnascope_vcf}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]: New line for each parameter

File: workflow/rules/sentieon.smk [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        threads=config.get("sentieon", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("sentieon", {}).get("time", config["default_resources"]["time"]),
+    container:
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Modify the dnascope vcf {input.vcf} by adding the MLrejected filter to the variants using Sentieon DNAModelApply"
+    shell:
+        "{params.sentieon} driver -t {threads} -r {params.reference} --algo DNAModelApply --model {params.model} -v {input.vcf} {output.vcf}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]: New line for each parameter

File: workflow/rules/sentieon.smk [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Call SNVs and structural variants in {input.tumorbam} using matched normal {input.normalbam} using Sentieon TNScope"
+    shell:
+        "{params.sentieon} driver -t {threads} -r {params.reference} "
+            "-i {input.tumorbam} -q {input.tumortable} -i {input.normalbam} -q {input.normaltable} "
+            "--algo TNscope --tumor_sample {wildcards.sample}_T --normal_sample {wildcards.sample}_N --bam_output {output.tnscope_bam} "
+            "{params.callsettings} {output.tnscope}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]: New line for each parameter

File: workflow/rules/sentieon.smk [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        time=config.get("sentieon", {}).get("time", config["default_resources"]["time"]),
+    container:
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Apply machine learning model on the TNScope vcf {input.tnscopevcf} to help with variant filtration using Sentieon TNModelApply"
+    shell:
+        "{params.sentieon} driver -t {threads} -r {params.reference} "
+        "--algo TNModelApply -m {params.model} -v {input.tnscopevcf} {output.vcf}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]: New line for each parameter

File: workflow/rules/sentieon.smk [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Align fastq files {input.reads} using Sentieon bwa mem against {params.reference}"
+    shell:
+        "{params.sentieon} bwa mem "
+            "-M -R '@RG\\tID:{wildcards.sample}_{wildcards.type}\\tSM:{wildcards.sample}_{wildcards.type}\\tPL:ILLUMINA' "
+            "-t {threads} {params.reference} {input.reads} "
+        "| {params.sentieon} util sort -o {output.bam} -t {threads} --sam2bam -i -"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]: New line for each parameter

File: workflow/scripts/dummy.py
Comment (jonca79) [inline workflow/scripts/dummy.py:1]: Keep dummy scripts for tests to function.

**Comment by jonca79** [comment]
The unita.tsv file in .tests/integration/ has missing columns (flowcell and barcode). That is what the dry-run error is.

File: workflow/scripts/dummy.py
Comment (hannasoderstrom1) [reply workflow/scripts/dummy.py:1]: restored in @ 833d26a

**Comment by jonca79** [comment]
You also need a .tests/integration/resources.yaml and a .tests/integration/config.yaml

File: workflow/rules/sentieon.smk (hunk 1 above)
Comment (hannasoderstrom1) [reply workflow/rules/sentieon.smk]: solved in @ 3cdec8a

File: workflow/Snakefile [hunk 8]
```diff
@@ -4,33 +4,8 @@ __email__ = "hanna.soderstrom@gu.se"
 __license__ = "GPL-3"
 
 include: "rules/common.smk"
```
Comment (jonca79) [inline workflow/Snakefile]:
```suggestion


include: "rules/common.smk"
```

File: workflow/rules/common.smk
Comment (jonca79) [inline workflow/rules/common.smk:1]: units = (
    pandas.read_table(config["units"], dtype=str)
    .set_index(["sample", "type", "flowcell", "lane", "barcode"], drop=False)
    .sort_index()
)

File: workflow/rules/common.smk [hunk 9]
```diff
@@ -42,9 +42,35 @@ wildcard_constraints:
...
+        "sentieon/qualcal/{}_{}_RECAL_DATA.TABLE".format(sample, t)
         for sample in get_samples(samples)
         for t in get_unit_types(units, sample)
     ]
+    output_files.append(
+        [
+            "sentieon/tnscope/{}_TNscope_tn_ML.vcf".format(sample)
+            for sample in get_samples(samples)
+        ]
+    )
```
Comment (jonca79) [inline workflow/rules/common.smk]:
```suggestion
    output_files.append(["sentieon/tnscope/{}_TNscope_tn_ML.vcf".format(sample) for sample in get_samples(samples)])
```

File: workflow/rules/sentieon.smk [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    params:
+        extra=config.get("sentieon", {}).get("extra", ""),
+        reference=config.get("sentieon", {}).get("reference", ""),
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+    log:
+        "sentieon/bwa_mem/{sample}_{flowcell}_{lane}_{barcode}_{type}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/bwa_mem/{sample}_{flowcell}_{lane}_{barcode}_{type}.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
            config.get("sentieon", {}).get("benchmark_repeats", 1),
```

File: workflow/rules/sentieon.smk [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+    container:
+        config.get("sentieon", {}).get("container", config["default_container"])
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Align fastq files {input.reads} using Sentieon bwa mem against {params.reference}"
+    shell:
+        "{params.sentieon} bwa mem "
+            "-M -R '@RG\\tID:{wildcards.sample}_{wildcards.type}\\tSM:{wildcards.sample}_{wildcards.type}\\tPL:ILLUMINA' "
+            "-t {threads} {params.reference} {input.reads} "
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} bwa mem "
        "-M -R '@RG\\tID:{wildcards.sample}_{wildcards.type}\\tSM:{wildcards.sample}_{wildcards.type}\\tPL:ILLUMINA' "
        "-t {threads} {params.reference} {input.reads} "
```

File: workflow/rules/sentieon.smk [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+        "| {params.sentieon} util sort -o {output.bam}"
+        "-t {threads}" 
+        "--sam2bam" 
+        "-i -"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "| {params.sentieon} util sort -o {output.bam} "
        "-t {threads} "
        "--sam2bam "
        "-i - &> {log}"
```

File: workflow/rules/sentieon.smk [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+    params:
+        extra=config.get("sentieon", {}).get("extra", ""),
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+    log:
+        "sentieon/dedup/{sample}_{type}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/dedup/{sample}_{type}.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
+        )
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        repeat("sentieon/dedup/{sample}_{type}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```

File: workflow/rules/sentieon.smk [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+            "--fun score_info "
+            "sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP_score.txt ;"
+        "{params.sentieon} driver "
+            "-t {threads} "
+            "-i {input} "
+            "--algo Dedup "
+            "--rmdup "
+            "--score_info sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP_score.txt "
+            "--metrics sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP.txt "
+            "sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP.bam"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-i {input} "
        "--algo LocusCollector "
        "--fun score_info "
        "sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP_score.txt ; "
        "{params.sentieon} driver "
        "-t {threads} "
        "-i {input} "
        "--algo Dedup "
        "--rmdup "
        "--score_info sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP_score.txt "
        "--metrics sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP.txt "
        "sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP.bam &> {log}"
```

File: workflow/rules/sentieon.smk [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        reference=config.get("sentieon", {}).get("reference", ""),
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+        mills=config.get("sentieon", {}).get("mills", ""),
+    log:
+        "sentieon/realign/{sample}_{type}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/realign/{sample}_{type}.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
+        )
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        repeat("sentieon/realign/{sample}_{type}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```

File: workflow/rules/sentieon.smk [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Indel realignment of bam file {input.bam} using Sentieon realigner"
+    shell:
+        "{params.sentieon} driver" 
+        "-t {threads}" 
+        "-r {params.reference}" 
+        "-i {input.bam}" 
+        "--algo Realigner" 
+        "-k {params.mills} {output}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver " 
        "-t {threads} " 
        "-r {params.reference} " 
        "-i {input.bam} " 
        "--algo Realigner " 
        "-k {params.mills} {output} &> {log}"
```

Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-r {params.reference} "
        "-i {input.bam} "
        "--algo Realigner "
        "-k {params.mills} {output} &> {log}"
```

File: workflow/rules/sentieon.smk [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+        mills=config.get("sentieon", {}).get("mills", ""),
+        dbsnp=config.get("sentieon", {}).get("dbsnp", ""),
+    log:
+        "sentieon/qualcal/{sample}_{type}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/qualcal/{sample}_{type}.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
+        )
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        repeat("sentieon/qualcal/{sample}_{type}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```

File: workflow/rules/sentieon.smk [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+    message:
+        "{rule}: Calculate recalibration table of {input.bam} using Sentieon QualCal algorithm"
+    shell:
+        "{params.sentieon} driver" 
+        "-t {threads}" 
+        "-r {params.reference}" 
+        "-i {input.bam}" 
+        "--algo QualCal" 
+        "-k {params.mills}" 
+        "-k {params.dbsnp} {output}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-r {params.reference} "
        "-i {input.bam} "
        "--algo QualCal "
        "-k {params.mills} "
        "-k {params.dbsnp} {output} &> {log}"
```

File: workflow/rules/sentieon.smk [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+        callsettings=config.get("sentieon", {}).get("dnascope_settings", ""),
+        model=config.get("sentieon", {}).get("dnascope_model", ""),
+        dbsnp=config.get("sentieon", {}).get("dbsnp", ""),
+    log:
+        "sentieon/dnascope/{sample}_{type}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/dnascope/{sample}_{type}.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
+        )
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        repeat(
            "sentieon/dnascope/{sample}_{type}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```

File: workflow/rules/sentieon.smk [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+        "{rule}: Call germline SNVs and structural variants in {input.bam} using Sentieon DNAScope"
+    shell:
+        "{params.sentieon} driver" 
+        "-t {threads}" 
+        "-r {params.reference} "
+            "-i {input.bam}" 
+            "--algo DNAscope" 
+            "-d {params.dbsnp} "
+            "--var_type snp,indel" 
+            "--model {params.model} {params.callsettings} {output.dnascope_vcf}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-r {params.reference} "
        "-i {input.bam} "
        "--algo DNAscope "
        "-d {params.dbsnp} "
        "--var_type snp,indel "
        "--model {params.model} {params.callsettings} {output.dnascope_vcf} &> {log}"
```

File: workflow/rules/sentieon.smk [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+        extra=config.get("sentieon", {}).get("extra", ""),
+        reference=config.get("sentieon", {}).get("reference", ""),
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+        model=config.get("sentieon", {}).get("dnascope_model", ""),
+    log:
+        "sentieon/dnascope/{sample}_{type}_modelfiter.output.log",
+    benchmark:
+        repeat(
+            "sentieon/dnascope/{sample}_{type}_modelfilter.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
            config.get("sentieon", {}).get("benchmark_repeats", 1),
```

File: workflow/rules/sentieon.smk [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+    conda:
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Modify the dnascope vcf {input.vcf} by adding the MLrejected filter to the variants using Sentieon DNAModelApply"
+    shell:
+        "{params.sentieon} driver -t {threads}" 
+        "-r {params.reference}" 
+        "--algo DNAModelApply" 
+        "--model {params.model}" 
+        "-v {input.vcf} {output.vcf}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver -t {threads} "
        "-r {params.reference} "
        "--algo DNAModelApply "
        "--model {params.model} "
        "-v {input.vcf} {output.vcf} &> {log}"
```

File: workflow/rules/sentieon.smk [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+        reference=config.get("sentieon", {}).get("reference", ""),
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+        callsettings=config.get("sentieon", {}).get("tnscope_settings", ""),
+    log:
+        "sentieon/tnscope/{sample}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/tnscope/{sample}.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
+        )
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        repeat("sentieon/tnscope/{sample}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```

File: workflow/rules/sentieon.smk [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+        "-r {params.reference} "
+            "-i {input.tumorbam}" 
+            "-q {input.tumortable}" 
+            "-i {input.normalbam}" 
+            "-q {input.normaltable} "
+            "--algo TNscope" 
+            "--tumor_sample {wildcards.sample}_T" 
+            "--normal_sample {wildcards.sample}_N" 
+            "--bam_output {output.tnscope_bam} "
+            "{params.callsettings} {output.tnscope}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-r {params.reference} "
        "-i {input.tumorbam} "
        "-q {input.tumortable} "
        "-i {input.normalbam} "
        "-q {input.normaltable} "
        "--algo TNscope "
        "--tumor_sample {wildcards.sample}_T "
        "--normal_sample {wildcards.sample}_N "
        "--bam_output {output.tnscope_bam} "
        "{params.callsettings} {output.tnscope} &> {log}"
```

File: workflow/rules/sentieon.smk [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+        callsettings=config.get("sentieon", {}).get("tnscope_settings", ""),
+        model=config.get("sentieon", {}).get("tnscope_model", ""),
+    log:
+        "sentieon/tnscope/{sample}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/tnscope/{sample}.output.benchmark.tsv",
+            config.get("sentieon", {}).get("benchmark_repeats", 1)
+        )
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        repeat("sentieon/tnscope/{sample}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```

File: workflow/rules/sentieon.smk [hunk 26]
```diff
@@ -0,0 +1,340 @@
...
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Apply machine learning model on the TNScope vcf {input.tnscopevcf} to help with variant filtration using Sentieon TNModelApply"
+    shell:
+        "{params.sentieon} driver" 
+        "-t {threads}" 
+        "-r {params.reference} "
+        "--algo TNModelApply" 
+        "-m {params.model}" 
+        "-v {input.tnscopevcf} {output.vcf}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-r {params.reference} "
        "--algo TNModelApply "
        "-m {params.model} "
        "-v {input.tnscopevcf} {output.vcf} &> {log}"
```

File: workflow/rules/common.smk
Comment (hannasoderstrom1) [reply workflow/rules/common.smk:1]: solved in @ f28cd91

File: workflow/rules/sentieon.smk (hunk 16 above)
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-r {params.reference} "
        "-i {input.bam} "
        "--algo Realigner "
        "-k {params.mills} {output} &> {log}"
```

File: workflow/rules/common.smk [hunk 27]
```diff
@@ -31,7 +31,11 @@ validate(samples, schema="../schemas/samples.schema.yaml")
 
 ### Read and validate units file
 
-units = pandas.read_table(config["units"], dtype=str).set_index(["sample", "type", "flowcell", "lane", "barcode"], drop=False).sort_index()
+units = (
+pandas.read_table(config["units"], dtype=str)
+.set_index(["sample", "type", "flowcell", "lane", "barcode"], drop=False)
+.sort_index()
+)
```
Comment (jonca79) [inline workflow/rules/common.smk]:
```suggestion
units = (
    pandas.read_table(config["units"], dtype=str)
    .set_index(["sample", "type", "flowcell", "lane", "barcode"], drop=False)
    .sort_index()
)
```

File: workflow/rules/sentieon.smk [hunk 28]
```diff
@@ -0,0 +1,323 @@
...
+    shell:
+        "{params.sentieon} bwa mem "
+        "-M -R '@RG\\tID:{wildcards.sample}_{wildcards.type}\\tSM:{wildcards.sample}_{wildcards.type}\\tPL:ILLUMINA' "
+        "-t {threads} {params.reference} {input.reads} "
+        "| {params.sentieon} util sort -o {output.bam} "
+        "-t {threads} "
+        "--sam2bam "
+        "-i - &> {log}"
+
+rule dedup:
```
Comment (jonca79) [inline workflow/rules/sentieon.smk:45]:
```suggestion


rule dedup:
```

File: workflow/rules/sentieon.smk (hunk 16 above)
Comment (hannasoderstrom1) [reply workflow/rules/sentieon.smk]: outdated, already updated

File: workflow/rules/sentieon.smk [hunk 29]
```diff
@@ -0,0 +1,323 @@
...
+        "{params.sentieon} driver "
+        "-t {threads} "
+        "-i {input} "
+        "--algo Dedup "
+        "--rmdup "
+        "--score_info sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP_score.txt "
+        "--metrics sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP.txt "
+        "sentieon/dedup/{wildcards.sample}_{wildcards.type}_DEDUP.bam &> {log}"
+
+rule realigner:
```
Comment (jonca79) [inline workflow/rules/sentieon.smk:90]:
```suggestion


rule realigner:
```

File: workflow/rules/sentieon.smk [hunk 30]
```diff
@@ -0,0 +1,323 @@
...
+        "{rule}: Indel realignment of bam file {input.bam} using Sentieon realigner"
+    shell:
+        "{params.sentieon} driver " 
+        "-t {threads} " 
+        "-r {params.reference} " 
+        "-i {input.bam} " 
+        "--algo Realigner " 
+        "-k {params.mills} {output} &> {log}"
+
+rule qualcal:
```
Comment (jonca79) [inline workflow/rules/sentieon.smk:126]:
```suggestion


rule qualcal:
```

File: workflow/rules/sentieon.smk [hunk 31]
```diff
@@ -0,0 +1,323 @@
...
+    shell:
+        "{params.sentieon} driver "
+        "-t {threads} "
+        "-r {params.reference} "
+        "-i {input.bam} "
+        "--algo QualCal "
+        "-k {params.mills} "
+        "-k {params.dbsnp} {output} &> {log}"
+
+rule dnascope:
```
Comment (jonca79) [inline workflow/rules/sentieon.smk:162]:
```suggestion


rule dnascope:
```

File: workflow/rules/sentieon.smk [hunk 32]
```diff
@@ -0,0 +1,323 @@
...
+        "{params.sentieon} driver "
+        "-t {threads} "
+        "-r {params.reference} "
+        "-i {input.bam} "
+        "--algo DNAscope "
+        "-d {params.dbsnp} "
+        "--var_type snp,indel "
+        "--model {params.model} {params.callsettings} {output.dnascope_vcf} &> {log}"
+
+rule dnascope_modelfilter:
```
Comment (jonca79) [inline workflow/rules/sentieon.smk:201]:
```suggestion


rule dnascope_modelfilter:
```

File: workflow/rules/sentieon.smk [hunk 33]
```diff
@@ -0,0 +1,323 @@
...
+    message:
+        "{rule}: Modify the dnascope vcf {input.vcf} by adding the MLrejected filter to the variants using Sentieon DNAModelApply"
+    shell:
+        "{params.sentieon} driver -t {threads} "
+        "-r {params.reference} "
+        "--algo DNAModelApply "
+        "--model {params.model} "
+        "-v {input.vcf} {output.vcf} &> {log}"
+
+rule tnscope:
```
Comment (jonca79) [inline workflow/rules/sentieon.smk:239]:
```suggestion


rule tnscope:
```

File: workflow/rules/sentieon.smk [hunk 34]
```diff
@@ -0,0 +1,323 @@
...
+        "-q {input.tumortable} "
+        "-i {input.normalbam} "
+        "-q {input.normaltable} "
+        "--algo TNscope "
+        "--tumor_sample {wildcards.sample}_T "
+        "--normal_sample {wildcards.sample}_N "
+        "--bam_output {output.tnscope_bam} "
+        "{params.callsettings} {output.tnscope} &> {log}"
+
+rule tnscope_modelfilter:
```
Comment (jonca79) [inline workflow/rules/sentieon.smk:283]:
```suggestion


rule tnscope_modelfilter:
```

File: workflow/rules/sentieon.smk [hunk 35]
```diff
@@ -0,0 +1,323 @@
...
+        reference=config.get("sentieon", {}).get("reference", ""),
+        sentieon=config.get("sentieon", {}).get("sentieon", ""),
+        callsettings=config.get("sentieon", {}).get("dnascope_settings", ""),
+        model=config.get("sentieon", {}).get("dnascope_model", ""),
+        dbsnp=config.get("sentieon", {}).get("dbsnp", ""),
+    log:
+        "sentieon/dnascope/{sample}_{type}.output.log",
+    benchmark:
+        repeat(
+            "sentieon/dnascope/{sample}_{type}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
    benchmark:
        repeat("sentieon/dnascope/{sample}_{type}.output.benchmark.tsv", config.get("sentieon", {}).get("benchmark_repeats", 1))
```

File: workflow/rules/sentieon.smk [hunk 36]
```diff
@@ -0,0 +1,329 @@
...
+        "../envs/sentieon.yaml"
+    message:
+        "{rule}: Indel realignment of bam file {input.bam} using Sentieon realigner"
+    shell:
+        "{params.sentieon} driver " 
+        "-t {threads} " 
+        "-r {params.reference} " 
+        "-i {input.bam} " 
+        "--algo Realigner " 
+        "-k {params.mills} {output} &> {log}"
```
Comment (jonca79) [inline workflow/rules/sentieon.smk]:
```suggestion
        "{params.sentieon} driver "
        "-t {threads} "
        "-r {params.reference} "
        "-i {input.bam} "
        "--algo Realigner "
        "-k {params.mills} {output} &> {log}"
```

**Review by elleira — APPROVED** [review]

**Pull request merged**